cs/test/BlittableLogCompactionTests.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using System.Linq;
+using FASTER.core;
+using System.IO;
+using NUnit.Framework;
+
+namespace FASTER.test
+{
+
+    [TestFixture]
+    internal class BlittableLogCompactionTests
+    {
+        private FasterKV<KeyStruct, ValueStruct, InputStruct, OutputStruct, Empty, Functions> fht;
+        private IDevice log;
+
+        [SetUp]
+        public void Setup()
+        {
+            log = Devices.CreateLogDevice(TestContext.CurrentContext.TestDirectory + "\\hlog4.log", deleteOnClose: true);
+            fht = new FasterKV<KeyStruct, ValueStruct, InputStruct, OutputStruct, Empty, Functions>
+                (1L << 20, new Functions(), new LogSettings { LogDevice = log, MemorySizeBits = 15, PageSizeBits = 7 });
+            fht.StartSession();
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            fht.StopSession();
+            fht.Dispose();
+            fht = null;
+            log.Close();
+        }
+
+        [Test]
+        public void BlittableLogCompactionTest1()
+        {
+            InputStruct input = default(InputStruct);
+
+            const int totalRecords = 2000;
+            var start = fht.Log.TailAddress;
+            long compactUntil = 0;
+
+            for (int i = 0; i < totalRecords; i++)
+            {
+                if (i == 1000)
+                    compactUntil = fht.Log.TailAddress;
+
+                var key1 = new KeyStruct { kfield1 = i, kfield2 = i + 1 };
+                var value = new ValueStruct { vfield1 = i, vfield2 = i + 1 };
+                fht.Upsert(ref key1, ref value, Empty.Default, 0);
+            }
+
+            fht.Log.Compact(compactUntil);
+            Assert.IsTrue(fht.Log.BeginAddress == compactUntil);
+
+            // Read 2000 keys - all should be present
+            for (int i = 0; i < totalRecords; i++)
+            {
+                OutputStruct output = default(OutputStruct);
+                var key1 = new KeyStruct { kfield1 = i, kfield2 = i + 1 };
+                var value = new ValueStruct { vfield1 = i, vfield2 = i + 1 };
+
+                var status = fht.Read(ref key1, ref input, ref output, Empty.Default, 0);
+                if (status == Status.PENDING)
+                    fht.CompletePending(true);
+                else
+                {
+                    Assert.IsTrue(status == Status.OK);
+                    Assert.IsTrue(output.value.vfield1 == value.vfield1);
+                    Assert.IsTrue(output.value.vfield2 == value.vfield2);
```

Comment:
Shouldn't these be outside of the "else" ? I.e., validate output for async Reads as well?

Reply:
Async read results are checked in the ReadCompletionCallback in FunctionsCompaction.